README
Module: kpass
Version: 0.5
Distribution file: pykpass-0.5.tar.gz
Author: Shumon Huque <shuque -at- isc.upenn.edu>

    kpass( username, password, service, host, kt_pathname )

EXAMPLE USAGE:

	from kpass import kpass, KpassError
        try:
            rc = kpass("frank", "bozo", "blah", None, "FILE:/etc/my.keytab")
        except KpassError, diag:
            print "Error: %s" % str(diag)
        else:
            if (rc == 1):
                print "Authentication success"
            else:
                print "Authentication failure"

DESCRIPTION:

This Python extension module provides a simple function called
kpass() to perform password verification using Kerberos 5. It is 
intended for use by applications that cannot use the Kerberos 
protocol natively, but need to authenticate users against the 
Kerberos database. If it must be run on a system that receives a 
username and password over the network, steps should be taken to 
ensure that these are passed to that system in a cryptographically 
secure manner.

kpass() attempts to validate a given user's Kerberos username and
password. It does this in the following manner: it first obtains a 
Kerberos ticket for the specified service for the given username and 
password from the Kerberos AS. And then attempts to decrypt the ticket 
using the key stored in the specified keytable file to verify the 
authenticity of the AS response. The python 'None' type can be passed 
as the 4th (host) argument to use the fully canonicalized primary 
hostname of the system that the function is executed on. The fifth 
argument can also be 'None' to use the system's default keytab file 
(usually "FILE:/etc/krb5.keytab").

Note that previous versions of this module obtained a TGT from the
AS and then subsequently used that to obtain the service ticket from
the TGS. Directly obtaining the service ticket from the AS saves a
round trip with the KDC and the associated cryptographic computations.

Local password verification doesn't need a replay cache, so kpass() 
by disables it. This speeds things up quite a bit, if you are invoking 
this function many times in quick succession.

kpass() returns 1 if password verification is successful, 0 if the 
username or password is incorrect and raises a custom exception of
"KpassError" if a system error is encountered.

kpass() relies on obtaining Kerberos realm and KDC information 
from the invoking environment. Typically it will get this from
the system's Kerberos configuration file (krb5.conf) and/or DNS
records. One quick way to override the default environment is to
create a custom krb5.conf file and set the pathname to this file
as the value of the KRB5_CONFIG environment variable.

This distribution is accompanied by a PGP signature. My PGP public
key can be obtained from <http://www.huque.com/~shuque/pgp/> or one
of the PGP key servers.

PRE-REQUISITES:

	- Python 2.x
	- MIT Kerberos V5 Release 1.3.x or newer (this package
	  seems to work with recent versions of Heimdal also,
	  but I have not tested this extensively).
	- Creation of an application service principal on the
	  Kerberos server for use by this function.
	- Storing this principal and it's associated key in
	  a local keytab file.

INSTALLATION:

1. Edit the file "setup.py" if necessary, to reflect the proper
   locations of the MIT Kerberos 5 libraries, include files, and
   other paraphernalia for your system.

2. Build the distribution:

	python setup.py build

3. Install it (probably as 'root' to install system wide):

	python setup.py install

   (If you just want to install the module in your home directory,
   use "python setup.py install --home $HOME", which will usually 
   put it in $HOME/lib/python/)

4. For testing the function, you'll need to edit the test/test1.py
   file appropriately for your environment (or write your own
   python code).

Shumon Huque
E-mail: <shuque -at- isc.upenn.edu>
Web: http://www.huque.com/~shuque/
University of Pennsylvania.

Copyright (c) 2005 - 2011, Shumon Huque. All rights reserved.  
This program is free software; you can redistribute it and/or modify 
it under the same terms as Python itself.


CHANGELOG
2011-06-13
pykpass version 0.5
Changed method of disabling the MIT Kerberos replay cache. The old
method set a flag in krb5_auth_con_setflags(), which does not appear
to work correctly in newer versions of MIT krb5. The new method
sets the environment variable KRB5RCACHETYPE to "none" - this needs
MIT Kerberos 1.4.x or newer to work. For earlier versions the replay
cache won't be disabled.

2007-04-30
pykpass version 0.4
Fixed a memory leak.

2007-04-09
pykpass version 0.3
Performance optimization: obtain service ticket directly from AS instead 
of via TGT. Also some minor clean. And no more testing for MIT Kerberos
versions < 1.3.

2005-08-17
pykpass version 0.2
Contains some minor fixes to work better with the Heimdal Kerberos
distribution.

2005-06-03
pykpass version 0.1
Python extension for Kerberos 5 password verification.


test/test1.py
#!/usr/local/bin/python
#

import kpass

USERNAME         = "foo"
PASSWORD_GOOD    = "spam"
PASSWORD_BAD     = "ham"
SERVICE          = "sname"
#SERVICE         = None
#HOSTNAME        = "blah.xyz.com"
HOSTNAME         = None
KT_PATHNAME      = "FILE:/path/to/my/file.keytab"
KT_BAD           = "FILE:/tmp/blah6873254834"

try:

    print ">>> Called with correct username/password:"
    print kpass.kpass(USERNAME, PASSWORD_GOOD, SERVICE, HOSTNAME, KT_PATHNAME)

    print ">>> Called with incorrect password:"
    print kpass.kpass(USERNAME, PASSWORD_BAD, SERVICE, HOSTNAME, KT_PATHNAME)

    print ">>> Called with bad keytab file:"
    print kpass.kpass(USERNAME, PASSWORD_GOOD, SERVICE, HOSTNAME, KT_BAD)

except kpass.KpassError, diag:

    print "Exception: %s: %s" % (diag.__class__.__name__, str(diag))
    


test/test2.py
#!/usr/local/bin/python
#

from kpass import kpass, KpassError

USERNAME         = "foo"
PASSWORD         = "spam"
SERVICE          = "sname"
#HOSTNAME        = "blah.xyz.com"
HOSTNAME         = None
KT_PATHNAME      = "FILE:/path/to/my/file.keytab"


try:
    rc = kpass(USERNAME, PASSWORD, SERVICE, HOSTNAME, KT_PATHNAME)
except KpassError, diag:
    print "Error: %s" % str(diag)
else:
    if (rc == 1):
        print "Authentication success"
    else:
        print "Authentication failure"
        


kpass.c
/*
 * kpass.c: Kerberos 5 Password Verification function
 *
 * Written by: Shumon Huque <shuque@upenn.edu>
 *
 */

#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <time.h>
#include <unistd.h>
#include <krb5.h>

#include "kpass.h"

#define TKT_LIFETIME     30             /* 30 seconds */
#define USE_MEMORY_CC                   /* use in-memory ccache */

static char rcsid[] = "$Id: kpass.c,v 1.5 2011/06/14 01:53:26 shuque Exp $";

/*
 * static variable to hold error message returned via exceptions
 */

static char kpass_error_message[256];

/*
 * this is used by the Python wrapper function (wrap_kpass) to 
 * obtain the Kerberos error message.
 */

char *obtain_errormsg(void) {
    return kpass_error_message;
}

/*
 * Check a Kerberos function's return value; if any error occurred, 
 * save the descriptive error message, which will be used later by
 * the module's Python exception class.
 */

#ifndef HEIMDAL
#define FAIL(err, tag) \
	if (err) { \
            (void) snprintf(kpass_error_message, 256, "%s %s", \
                            error_message(err), tag); \
	    goto cleanup; \
	}
#else                     /* Heimdal: use krb5_get_err_text */
#define FAIL(err, tag) \
	if (err) { \
            (void) snprintf(kpass_error_message, 256, "%s %s", \
                            krb5_get_err_text(context, err), tag); \
	    goto cleanup; \
	}
#endif

/*
 * Kerberos password verification. Attempt to obtain credentials for a given
 * username and password from Kerberos AS, then obtain credentials for a local
 * service from Kerberos TGS to verify the authenticity of the AS response.
 *
 * Arguments:
 *	username, password	The username and password to verify.
 *	service, host		A local service whose key will be used to
 *                              verify the authenticity of the Kerberos 
 *                              credentials. Specifying host as NULL
 *                              says use my primary hostname.
 *	kt_pathname		Path of the keytab file, or NULL to use default
 *				keytab. (Typical form "FILE:/abs/path/name".)
 *
 * Return -1 if an error occurs, 0 if the username or password is incorrect,
 * or 1 if password verification is successful.
 */


int kpass(username, password, service, host, kt_pathname)
     char *username;
     char *password;
     char *service;
     char *host;
     char *kt_pathname;
{
    krb5_error_code               err;
    krb5_context                  context;
    krb5_auth_context             auth_context = NULL;
    krb5_creds                    credentials;
    krb5_principal                user_principal,
	                          service_principal;
    krb5_keytab                   keytab = NULL;
    krb5_ccache                   ccache;
    char                          ccache_name[L_tmpnam + 8];
    krb5_get_init_creds_opt       gic_options;
    krb5_verify_init_creds_opt    vic_options;
    krb5_data                     apreq_pkt;
    char                          myhostname[256], sprinc[256];

    int                           have_ccache = 0,
	                          have_user_principal = 0,
	                          have_service_principal = 0,
	                          have_keytab = 0,
	                          have_credentials = 0,
                                  success = -1;

    apreq_pkt.data = NULL;

    err = krb5_init_context(&context);
    FAIL(err, "from krb5_init_context");

#ifdef NEED_INIT_ETS
    krb5_init_ets(context);
#endif

#ifndef HEIMDAL
    /* Disable replay cache.  Requires MIT Kerberos 1.4.x or later */
    if (getenv("KRB5RCACHETYPE") == NULL)
        putenv("KRB5RCACHETYPE=none");
#endif

#ifdef USE_MEMORY_CC
    (void) memset(ccache_name, 0, sizeof(ccache_name));
    (void) strcpy(ccache_name, "MEMORY:");
    (void) tmpnam(&ccache_name[7]);
    err = krb5_cc_resolve(context, ccache_name, &ccache);
    FAIL(err, "from krb5_cc_resolve");
#else
    err = krb5_cc_default(context, &ccache);
    FAIL(err, "from cc_default");
#endif
    have_ccache = 1;

    err = krb5_parse_name(context, username, &user_principal);
    FAIL(err, "from krb_parse_name");
    have_user_principal = 1;

    err = krb5_cc_initialize(context, ccache, user_principal);
    FAIL(err, "from krb_cc_initialize");

    (void) memset( (char *)&credentials, 0, sizeof(credentials) );
    if (!host) {
        (void) gethostname(myhostname, sizeof(myhostname));
        snprintf(sprinc, sizeof(sprinc), "%s/%s", service, myhostname);
    } else
        snprintf(sprinc, sizeof(sprinc), "%s/%s", service, host);

    krb5_get_init_creds_opt_init(&gic_options);
    krb5_get_init_creds_opt_set_tkt_life(&gic_options, TKT_LIFETIME);

    err = krb5_get_init_creds_password(context, &credentials,
                                       user_principal, password,
                                       0, 0, 0, sprinc, &gic_options);

    switch (err) {
    case 0:
	/* Success. */
	have_credentials = 1;
	break;
    case KRB5KDC_ERR_PREAUTH_FAILED:
    case KRB5KRB_AP_ERR_BAD_INTEGRITY:
    case KRB5KDC_ERR_C_PRINCIPAL_UNKNOWN:
	/* Bad username or password, unknown user etc */
	success = 0;
	goto cleanup;
    default:
	/* Some other sort of failure. */
	FAIL(err, "from krb5_get_init_creds_password");
	break;
    }

    err = krb5_cc_store_cred(context, ccache, &credentials);
    FAIL(err, "from krb5_cc_store_cred");

    err = krb5_sname_to_principal(context, host, service,
				  KRB5_NT_SRV_HST, &service_principal);
    FAIL(err, "from krb5_sname_to_principal");
    have_service_principal = 1;

    if (kt_pathname && *kt_pathname) {
	err = krb5_kt_resolve(context, kt_pathname, &keytab);
	FAIL(err, "from krb5_kt_resolve");
	have_keytab = 1;
    }

    krb5_verify_init_creds_opt_init(&vic_options);
    krb5_verify_init_creds_opt_set_ap_req_nofail(&vic_options, 1);
    err = krb5_verify_init_creds(context, &credentials, service_principal,
                                 keytab, 0, &vic_options);
    FAIL(err, "from krb5_verify_init_creds");

    success = 1;

cleanup:

    if (have_ccache)
        krb5_cc_destroy(context, ccache);

    if (apreq_pkt.data)
	krb5_free_data_contents(context, &apreq_pkt);

    if (have_keytab)
	if ( (err = krb5_kt_close(context, keytab)) )
	    com_err("kpass", err, "from krb5_kt_close");

    if (have_user_principal)
	krb5_free_principal(context, user_principal);

    if (have_service_principal)
	krb5_free_principal(context, service_principal);

    if (have_credentials)
	krb5_free_cred_contents(context, &credentials);

    if (auth_context) {
     	krb5_auth_con_free(context, auth_context);
	auth_context = NULL;
    }

    if (context)
	krb5_free_context(context);

    return success;
}



kpass.py
"""
Python extension for Kerberos 5 password verification.

This Python extension module provides a simple function called
kpass() to perform password verification using Kerberos 5. It is 
intended for use by applications that cannot use the Kerberos 
protocol natively, but need to authenticate users against the 
Kerberos database. If it must be run on a system that receives a 
username and password over the network, steps should be taken to 
ensure that these are passed to that system in a cryptographically
secure manner.

kpass() attempts to obtain credentials for the given username and 
password from the Kerberos AS, then obtain credentials for a local 
service from the Kerberos TGS to verify the authenticity of the AS 
response. The python 'None' type can be passed as the 3rd (service)
and 4th (host) arguments to use the default service name (host) 
and the fully canonicalized primary hostname of the system that the 
function is executed on. The fifth argument can also be 'None' to
use the system's default keytab file (usually FILE:/etc/krb5.keytab).

kpass() returns 1 if password verification is successful, 0 if the 
username or password is incorrect and raises a custom exception of
KpassError if a system error is encountered.

kpass() relies on obtaining Kerberos realm and KDC information 
from the invoking environment. Typically it will get this from
the system's Kerberos configuration file (krb5.conf) and/or DNS
records. One quick way to override the default environment is to
create a custom krb5.conf file and setting the pathname to this 
file as the value of the KRB5_CONFIG environment variable.
"""

import _kpass

VERSION = "0.5"

KpassError = _kpass.KpassError

def kpass(*args):
    """kpass(username, password, service, host, kt_pathname) -> int"""
    return _kpass.kpass(*args)



wrap_kpass.c
#include <Python.h>
#include <stdlib.h>
#include "kpass.h"

static char rcsid[] = "$Id: wrap_kpass.c,v 1.2 2007/04/10 03:02:00 shuque Exp $";

static PyObject *KpassError;
 
static PyObject *
wrap_kpass(PyObject *self, PyObject *args) {

    char *username, *password, *service, *host, *kt_pathname;
    int  rc;

    if ( !PyArg_ParseTuple(args, "ssszz", 
			   &username, &password, &service, &host, 
			   &kt_pathname) )
	return NULL;

    rc = kpass(username, password, service, host, kt_pathname);

    if (rc == -1) {
	PyErr_SetString(KpassError, obtain_errormsg());
	return NULL;
    }

    return Py_BuildValue("i", rc);

}
 
static struct PyMethodDef kpass_methods[] = {
    {"kpass", wrap_kpass, METH_VARARGS,  "Kerberos 5 password verification"},
    {NULL, NULL, 0, NULL}
};
 
 
/*
 * Initialization: called by Python on first import
 */

PyMODINIT_FUNC
init_kpass() {

    PyObject *m;

    m = Py_InitModule("_kpass", kpass_methods);
    KpassError = PyErr_NewException("kpass.KpassError", 
				    PyExc_Exception, NULL);
    Py_INCREF(KpassError);
    PyModule_AddObject(m, "KpassError", KpassError);

}
